### PR TITLE
Chapter 9: Advanced login

### PR DESCRIPTION
**Pull request by phucynwa**

### CODE REVIEW

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -3,15 +3,30 @@ def log_in user
     session[:user_id] = user.id
   end
 
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: I have a question: What is signed cookie, I meant what is the purpose of using signed cookie here?

Comment (phucynwa) [reply app/helpers/sessions_helper.rb]: @mdmahmudurrahman cookie keep login infomation in client to help server detect client has logged in before when client visit website.

Comment (mdmahmudurrahman) [reply app/helpers/sessions_helper.rb]: @phucynwa No, I meant why you used signed here?

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -3,15 +3,30 @@ def log_in user
...
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
+      @current_user ||= User.find_by_id user_id
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: It is not recommended to use dynamic finder(find_by_id) now. You can use find_by here.

File: app/models/user.rb [hunk 3]
```diff
@@ -13,5 +14,23 @@ def digest string
...
       BCrypt::Password.create string, cost: cost
     end
+
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+
+  def remember
+    self.remember_token = User.new_token
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: `remember_token = User.new_token`

File: app/models/user.rb [hunk 4]
```diff
@@ -13,5 +14,23 @@ def digest string
...
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. Use update_attributes
2. Can study the question: What is the difference between update_attributes and update_attribute

File: app/helpers/sessions_helper.rb (hunk 1 above)
Comment (phucynwa) [reply app/helpers/sessions_helper.rb]: @mdmahmudurrahman Yeah, it was used in tut, but I think it isn't really need in this case. User ID haven't to be encode in this case.

File: app/models/user.rb [hunk 5]
```diff
@@ -13,5 +14,23 @@ def digest string
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: remove last ()

File: app/models/user.rb [hunk 6]
```diff
@@ -13,5 +14,23 @@ def digest string
...
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Use update_attributes

File: app/helpers/sessions_helper.rb (hunk 1 above)
Comment (mdmahmudurrahman) [reply app/helpers/sessions_helper.rb]: No, you study about signed cookie first, it's purpose, usefulness etc.

Comment (phucynwa) [reply app/helpers/sessions_helper.rb]: @mdmahmudurrahman it prevents users change their user id. right ?

Comment (mdmahmudurrahman) [reply app/helpers/sessions_helper.rb]: @phucynwa And this user is a malicious user/hacker who temper cookie.  =))

File: app/models/user.rb [hunk 7]
```diff
@@ -13,5 +14,23 @@ def digest string
...
     end
+
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+
+  def remember
+    remember_token = User.new_token
+    update_attributes :remember_digest => User.digest(remember_token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Convention: Hash racket syntax is deprecated in rails, please use symbol

File: app/models/user.rb [hunk 8]
```diff
@@ -13,5 +14,23 @@ def digest string
...
+  end
+
+  def authenticated? remember_token
+    return false if remember_digest.nil?
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attributes :remember_digest => nil
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Same here.

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -3,15 +3,30 @@ def log_in user
...
 
+  def remember user
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if user_id = session[:user_id]
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: Is it correct? Can you check it?

**Pull request merged**